### ISSUE TITLE
Compilation failure with modules that mix different compilation options

### ISSUE THREAD

**Issue by hugoam** [llvm/llvm-project#53231]
The following simple modules based program compiles correctly with the commands below, however as soon as you add some different compilation options in the module and the consumer code things start to go haywire.
Examples of options that will cause a failure if you add it to the first command but not the second:
- Any macro define like `-DMYMACRO`
- `-fPIC`
- Probably many many others (I haven't had time to check each one), I assume each `-f` switch, etc.

greetings.cpp
```
export module greetings;
 
import <string>;
 
export std::string get_greeting_text()
{
    return "Hello, World!";
}
```

main.cpp
```
import <iostream>;
import greetings;

int main()
{
    std::cout << get_greeting_text() << '\n';
}
```

Compile commands:
```
clang++ -Wall -std=c++2a -stdlib=libc++ -fimplicit-modules -fimplicit-module-maps -fmodules --precompile -x c++-module greetings.cpp -o greetings.pcm
clang++ -fmodules -c greetings.pcm -o greetings.o
clang++ -Wall -std=c++2a -stdlib=libc++ -fimplicit-modules -fimplicit-module-maps -fprebuilt-module-path=. greetings.o main.cpp -o main
```

Failing commands:
```
clang++ -Wall -std=c++2a -stdlib=libc++ -fimplicit-modules -fimplicit-module-maps -fmodules --precompile -x c++-module -DMYMACRO greetings.cpp -o greetings.pcm
clang++ -fmodules -c greetings.pcm -o greetings.o
clang++ -Wall -std=c++2a -stdlib=libc++ -fimplicit-modules -fimplicit-module-maps -fprebuilt-module-path=. greetings.o main.cpp -o main
```

The error produced is systematically a variation of:
```
main.cpp:2:1: fatal error: module 'std_config' is defined in both '/home/hamiard/.cache/clang/ModuleCache/1LVJ3VD5M3URQ/std_config-2FRLOTTTWC8AB.pcm' and '/home/hamiard/.cache/clang/ModuleCache/28LE54LT49FD8/std_config-2FRLOTTTWC8AB.pcm'
```

I found that clang has an option `-fmodules-ignore-macro` which can help alleviate the problem in the specific case of macros, but this is technically duct tape, it really doesn't address the root issue. Mixing modules with different compile options should be totally fine except for a very few exceptional cases, shouldn't it ?

The other compilers seem handle such benign differences fine (except for a few critical macros) so I believe clang should as well. Not to mention this practically makes building large modules based projects almost strictly impossible. Is there as hidden setting or subtlety I yet don't know of that can help solve this problem or is this a bug ?

**Comment by torokati44** [comment]
While trying to slowly (class-by-class) modularize a sizable C++ project, after hitting a really similar (honestly confusing) error message _(just with `_Builtin_stddef_max_align_t` in place of `std_config`)_, I've arrived to an almost identical reproducer to what's in the OP.
So I'd like to add my +1 as a friendly ping to this issue.

**Comment by ChuanqiXu9** [comment]
> While trying to slowly (class-by-class) modularize a sizable C++ project, after hitting a really similar (honestly confusing) error message _(just with `_Builtin_stddef_max_align_t` in place of `std_config`)_, I've arrived to an almost identical reproducer to what's in the OP. So I'd like to add my +1 as a friendly ping to this issue.

I got the reason why this issue is solved slowly. It is because it comes from a mixed use of C++20 modules and the clang modules. And we're still unclear what to do for the mixed use. See https://clang.llvm.org/docs/StandardCPlusPlusModules.html#introduction and https://clang.llvm.org/docs/StandardCPlusPlusModules.html#interoperability-with-clang-modules

**Comment by torokati44** [comment]
> it comes from a mixed use of C++20 modules and the clang modules

I see, thank you for the explanation! I suppose "Clang modules" here are used implicitly somehow...? Would those be difficult to disable in these cases, or just not worth it?
Anyway, I too was able to get past this error using `-fmodules-ignore-macro=FOOBAR` as suggested above.

> And we're still unclear what to do for the mixed use.

In the meantime, I think just detecting this case and changing the error message to something that more directly addresses the cause (even listing exactly the problematic/differing macros perhaps), and maybe even mentions this potential workaround, would be a significant improvement.

**Comment by ChuanqiXu9** [comment]
> Would those be difficult to disable in these cases, or just not worth it?

No, it is easy and I tried before. But the existing users of clang modules expressed the intention to adopt C++20 modules step by step instead of a big change at once. So it is not good to emit an error directly if we detected the mixed use of clang modules and C++20 modules.

The take away here may be to read the documents to figure out where the line is. The options listed in https://clang.llvm.org/docs/StandardCPlusPlusModules.html are intended to be options to support C++20 modules. While it is possible that there are more options from clang modules side that works in C++20 modules, it is undefined whether or not these options works for C++20 modules formally.

**Comment by torokati44** [comment]
> But the existing users of clang modules expressed the intention to adopt C++20 modules step by step instead of a big change at once.

I meant on a case-by-case basis, for a single project.

> ... to figure out where the line is.

The line, as in, what to disable, and what not to?

**Comment by ChuanqiXu9** [comment]
> The line, as in, what to disable, and what not to?

If you only want to use C++20 modules, you should disable all the use of module flags not mentioned in https://clang.llvm.org/docs/StandardCPlusPlusModules.html .

**Comment by torokati44** [comment]
Okay, I just realized, I had fallen victim to a grave misunderstanding.
For who knows what reason, I thought that `-std=c++20` was not enough to enable standard C++ modules in Clang, and I also needed `-fmodules` for that. In fact, this does not appear to be the case. :facepalm: `-fmodules` enables the other, Clang-modules mechanism, which caused the clash, and the error in my case.

So, simply removing `-fmodules` also solved my issue in both my reduced reproducer, and in the original project.

**Comment by ChuanqiXu9** [comment]
Yes, and ideally the users don't need to fight with these flags. They should depend on the build systems directly by the design.

**Comment by torokati44** [comment]
Sure, in an ideal (or future) world, where most build systems have working support for C++ modules out of the box. :)
Until then, some of us have to make do with our custom build machinery, smelling distinctly like the '90s.

**Comment by ChuanqiXu9** [comment]
BTW, cmake and xmake has already supported named modules. I think you can take a try on them if you're not a build system writer.

**Comment by torokati44** [comment]
Yeah, I know about those, thanks! Unfortunately we are currently stuck with our own setup, but have been considering throwing it out in favor of something more modern for a while now.
